pkg/datapath/probes/probes.go

```diff
@@ -0,0 +1,347 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package probes
+
+import (
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/multierror"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "probes")
+
+	localConfigLocations = []string{
+		"/proc/config",
+		fmt.Sprintf("/boot/config-%s", getKernelVersionStr()),
+	}
+	localConfigLocationsGz = []string{
+		"/proc/config.gz",
+	}
+)
+
+// fileWithCleanup is a wrapper for os.File which removes the file on closing
+// when no bytes were written.
+type fileWithCleanup struct {
+	path string
+	n    int
+	*os.File
+}
+
+// Close closes the File, rendering it unusable for I/O. On files that support
+// SetDeadline, any pending I/O operations will be canceled and return
+// immediately with an error. If no bytes were written to the file, it will be
+// removed after closing.
+func (f *fileWithCleanup) Close() error {
+	if err := f.File.Close(); err != nil {
+		return fmt.Errorf("could not close file %s: %s", f.path, err)
+	}
+	if f.n > 0 {
+		if err := os.Remove(f.path); err != nil && !os.IsNotExist(err) {
+			return fmt.Errorf("could not remove file %s: %s", f.path, err)
+		}
+	}
+	return nil
+}
+
+// Sync commits the current contents of the file to stable storage. Typically,
+// this means flushing the file system's in-memory copy of recently written
+// data to disk.
+func (f *fileWithCleanup) Sync() error {
```

Comment:
Can you just return f.File.Sync here? it doesn't seem like this implementation is doing anything else.

Reply:
I just wanted to extend the error message. I think it's a bit cleaner than writing

```
if err != nil {
    return fmt.Errorf("could not sync file %s: %s", f.path, err)
}
```

multiple times. With that override, I can just do `if err != nil { return err }` later.